// @adeira/css-colors/src/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "calculateContrastRatio", {
  enumerable: true,
  get: function get() {
    return _calculateContrastRatio.default;
  }
});
Object.defineProperty(exports, "chooseHigherContrast", {
  enumerable: true,
  get: function get() {
    return _chooseHigherContrast.default;
  }
});
Object.defineProperty(exports, "convertToRGBTriplet", {
  enumerable: true,
  get: function get() {
    return _convertToRGBTriplet.default;
  }
});
Object.defineProperty(exports, "cssColorNames", {
  enumerable: true,
  get: function get() {
    return _cssColorNames.default;
  }
});
Object.defineProperty(exports, "hex3ToHex6", {
  enumerable: true,
  get: function get() {
    return _hex3ToHex.default;
  }
});
Object.defineProperty(exports, "hex6ToHex3", {
  enumerable: true,
  get: function get() {
    return _hex6ToHex.default;
  }
});
Object.defineProperty(exports, "isAccessible", {
  enumerable: true,
  get: function get() {
    return _isAccessible.default;
  }
});
Object.defineProperty(exports, "isBright", {
  enumerable: true,
  get: function get() {
    return _isBright.default;
  }
});
Object.defineProperty(exports, "isColor", {
  enumerable: true,
  get: function get() {
    return _isColor.default;
  }
});
Object.defineProperty(exports, "isDark", {
  enumerable: true,
  get: function get() {
    return _isDark.default;
  }
});
Object.defineProperty(exports, "normalizeColor", {
  enumerable: true,
  get: function get() {
    return _normalizeColor.default;
  }
});

var _calculateContrastRatio = _interopRequireDefault(require("./calculateContrastRatio"));

var _chooseHigherContrast = _interopRequireDefault(require("./chooseHigherContrast"));

var _convertToRGBTriplet = _interopRequireDefault(require("./convertToRGBTriplet"));

var _cssColorNames = _interopRequireDefault(require("./cssColorNames"));

var _hex3ToHex = _interopRequireDefault(require("./hex3ToHex6"));

var _hex6ToHex = _interopRequireDefault(require("./hex6ToHex3"));

var _isAccessible = _interopRequireDefault(require("./isAccessible"));

var _isBright = _interopRequireDefault(require("./isBright"));

var _isColor = _interopRequireDefault(require("./isColor"));

var _isDark = _interopRequireDefault(require("./isDark"));

var _normalizeColor = _interopRequireDefault(require("./normalizeColor"));
